File: client/javascript/src/prsim/builtin.ts
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */

import {spi} from "@/macro/spi";
import {mpi} from "@/macro/mpi";
import type {Context} from "@/prsim/context";
import {Versions} from "@/kinds/versions";
import {Status} from "@/cause";
import {index} from "@/macro";

@spi("mesh")
export abstract class Builtin {

    /**
     * Doc export the documents.
     */
    @mpi("${mesh.name}.builtin.doc", String)
    doc(ctx: Context, @index(0, 'name') name: string, @index(1, 'formatter') formatter: string): Promise<string> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    /**
     * Version will get the builtin application version.
     */
    @mpi("${mesh.name}.builtin.version", Versions)
    version(ctx: Context): Promise<Versions> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    /**
     * Debug set the application log level.
     */
    @mpi("${mesh.name}.builtin.debug")
    debug(ctx: Context, @index(0, 'features') features: Map<string, string>): Promise<void> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    /**
     * Stats will collect health check stats.
     */
    @mpi("${mesh.name}.builtin.stats", [Map, String, String])
    stats(ctx: Context, @index(0, 'features') features: string[]): Promise<Map<string, string>> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    /**
     * Fallback is fallback service
     */
    @mpi("${mesh.name}.builtin.fallback")
    fallback(ctx: Context): Promise<void> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }
}

File: client/golang/prsim/builtin.go
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */

package prsim

import (
	"context"
	"github.com/opendatav/mesh/client/golang/types"
	"runtime"
)

var (
	Version  = "1.5.0.0"
	CommitID = "59a06ccd"
	GOOS     = runtime.GOOS
	GOARCH   = runtime.GOARCH
)

var IBuiltin = (*Builtin)(nil)

// Builtin spi
// @MPI(macro.MeshMPI)
// @SPI(macro.MeshSPI)
type Builtin interface {

	// Doc export the documents.
	// @MPI("${mesh.name}.builtin.doc")
	Doc(ctx context.Context, name string, formatter string) (string, error)

	// Version will get the builtin application version.
	// @MPI("${mesh.name}.builtin.version")
	Version(ctx context.Context) (*types.Versions, error)

	// Debug set the application log level.
	// @MPI("${mesh.name}.builtin.debug")
	Debug(ctx context.Context, features map[string]string) error

	// Stats will collect health check stats.
	// @MPI("${mesh.name}.builtin.stats")
	Stats(ctx context.Context, features []string) (map[string]string, error)

	// Fallback is fallback service
	// @MPI("${mesh.name}.builtin.fallback")
	Fallback(ctx context.Context) error
}

var IHodor = (*Hodor)(nil)

type Hodor interface {

	// X only diff with Builtin
	X() string

	// Stats collect the system, application, process or thread status etc.
	Stats(ctx context.Context, features []string) (map[string]string, error)

	// Debug set the debug features.
	Debug(ctx context.Context, features map[string]string) error
}
